### ISSUE TITLE
"Make reference types nullable" must use nullable type in SetRelatedEntity()

### ISSUE THREAD

**Issue by randombyte-developer** [daryllabar/DLaB.Xrm.XrmToolBoxTools#501]
**Tool and Version**
EarlyBoundGenerator V 2.2024.5.16

**Describe the bug**
I wanted to try out https://github.com/daryllabar/DLaB.Xrm.XrmToolBoxTools/issues/311. It generates nullable types. When compiling the generated code there is a compile error:

```
Possible null reference argument for parameter 'entity' in 'void Entity.SetRelatedEntity<Contact>(string relationshipSchemaName, EntityRole? primaryEntityRole, Contact entity)'
```

The relevant line of code is
`this.SetRelatedEntity<Contact>("account_primary_contact", null, value);`

`value` can be null in this case. SetRelatedEntity() does handle null `value`s correctly but the nullable entity type has to be used like this:
`this.SetRelatedEntity<Contact?>("account_primary_contact", null, value);`

**To Reproduce**
1. Generate the code for an Account with nullable types enabled.
2. Compile it.

**Expected behavior**
No errors.

**Additional context**
My C# project file looks like this:

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
		<SignAssembly>true</SignAssembly>
		<AssemblyOriginatorKeyFile>........snk</AssemblyOriginatorKeyFile>
	</PropertyGroup>
	<PropertyGroup>
		<LangVersion>11.0</LangVersion>
		<TargetFramework>net462</TargetFramework>
		<Nullable>enable</Nullable>
		<Version>1.0.0.0</Version>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
	</PropertyGroup>
	<ItemGroup>
	  <PackageReference Include="Microsoft.CrmSdk.CoreAssemblies" Version="9.0.2.52" />
	</ItemGroup>
</Project>
```

**Comment by randombyte-developer** [comment]
A workaround is to not enable the `INotifyPropertyChanging` pattern in the EBG config.

**Comment by daryllabar** [comment]
Looks like with the INotifyPropertyChanging, also need to null out the PropertyChanged and Property Changing events:

```cs
public event System.ComponentModel.PropertyChangedEventHandler? PropertyChanged;
		
public event System.ComponentModel.PropertyChangingEventHandler? PropertyChanging;
```

**Comment by SvenRahnSIT** [comment]
@daryllabar Well this makes the code compile but I guess the correct way to fix this would have been the `?` as I wrote in the issue description. Technically right now it is wrong but it works in the context of the current implementation of `SetRelatedEntity()`.

**Comment by daryllabar** [comment]
Why is it technically wrong?

**Comment by SvenRahnSIT** [comment]
The `value` in the property setter can be null. By using `this.SetRelatedEntity<Contact?>("account_primary_contact", null, value);` (with the `?`) the compiler is fine because the whole property is nullable. But the null forgiving operator `!` should only be used when you are sure it is not null or explicitly want to pass null like in https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving#examples

Right now it doesn't make a difference but the correct choice would be to just use the nullable type.

**Comment by daryllabar** [comment]
Ah, so you're saying that because I know the function can handle null, my use of the null forgiving operator is incorrect, because that should only be used when you know it isn't null, not when you know the calling function can accept the null?

**Comment by SvenRahnSIT** [comment]
Right. It suppresses nullable warnings which is not necessary here because the function accepts a nullable type e.g. `Account?` so you don't need to suppress any warning.
